### PR TITLE
Login usecase

### PR DESCRIPTION
**Pull request by Mohan29997**
login page - make use-case for login api - flow from presentation layer inside it saga and then domain layer (usecase) data layer (implementation of use case and port out for network and database)and network layer and storing that data in database in watermelon db  and storing in reducer 
dashboard page - make use-case to get the data from database and showing in ui with a same flow of use-case then implementation and then port and database adapter

### CODE REVIEW

File: packages/data/src/out/database_port.ts [hunk 1]
```diff
@@ -1,5 +1,6 @@
-import { MyDatabaseModel, User } from 'packages/shared/src/shared'
+import { UserDetailsModel } from './../../../shared/src/model/userDetailsmodal'
+import { UserCheckModal } from 'shared'
```
Comment (naikdp7) [inline packages/data/src/out/database_port.ts:2]: rename to UserCheckModel

File: packages/mobile/src/feature/login/login_screen.tsx [hunk 2]
```diff
@@ -7,23 +8,37 @@ import { useDispatch, useSelector } from 'react-redux'
...
-    dispatch(userRequest(data))
+    if (username == '' || password == '') {
+      alert(i18n.t('noInput'))
+    } else {
+      dispatch(userRequest({ data: data }))
+    }
   }
+
+  useEffect(() => {
+    if (loginData?.status == 1) setLodingState(true)
```
Comment (naikdp7) [inline packages/mobile/src/feature/login/login_screen.tsx:34]: can we create a enum/constant to define the statuses?

**Review by naikdp7 — CHANGES_REQUESTED** [review]
Requested changes

File: packages/mobile/src/feature/login/login_screen.tsx (hunk 2 above)
Comment (Mohan29997) [reply packages/mobile/src/feature/login/login_screen.tsx]: created enum for same in utils folder inside packages/mobile/utils/status.ts

File: packages/data/src/out/network_port.ts [hunk 3]
```diff
@@ -1,5 +1,6 @@
-import { MyFirstModel } from 'packages/shared/src/shared'
+import { MyFirstModel, UserCheckModel } from 'shared'
```
Comment (faiyyazs) [inline packages/data/src/out/network_port.ts:1]: Remove unused code

File: packages/database-watermelon/src/safe_db_call.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+async function SafeDbCall<T>(dbcall: Promise<T>): Promise<T> {
```
Comment (faiyyazs) [inline packages/database-watermelon/src/safe_db_call.ts:1]: make it safeDbCall(..)

File: packages/database-watermelon/src/dao/user_dao.ts [hunk 5]
```diff
@@ -1,20 +1,40 @@
...
+  readonly tablename: string
+  readonly databaseData: Collection<UserModel>
+
+  tableName(): string {
     return UserModel.table
   }
 
   constructor(appdatabase: AppDatabase) {
     super(appdatabase)
+    this.tablename = this.tableName()
```
Comment (faiyyazs) [inline packages/database-watermelon/src/dao/user_dao.ts:19]: remove this initialisation

File: packages/database-watermelon/src/dao/user_dao.ts [hunk 6]
```diff
@@ -1,20 +1,40 @@
...
 import { AppDatabase } from './../app_database'
 import { BaseDao } from './base_dao'
-import { Q } from '@nozbe/watermelondb'
+import { Collection, Q } from '@nozbe/watermelondb'
+import SafeDbCall from '../safe_db_call'
 
 export class UserDao extends BaseDao<UserModel> {
-  tableName() {
+  readonly tablename: string
+  readonly databaseData: Collection<UserModel>
```
Comment (faiyyazs) [inline packages/database-watermelon/src/dao/user_dao.ts:11]: not required

File: packages/data/src/repositories/user_repository_impl.ts [hunk 7]
```diff
@@ -1,19 +1,34 @@
...
+    this.database = params.databasePort
     this.network = params.networkPort
-    console.log('user_repo_epl')
   }
-  async loginCheck(params?: { userName: string }): Promise<boolean> {
-    return true
-    //return this.database.userCheck({ email: params.userName })
+
+  async loginCheck(params?: { username: string; password: string }): Promise<UserCheckModel> {
+    const message = 'Request failed with status code 403'
```
[the remark this answers is not in the document]
Comment (Mohan29997) [reply packages/data/src/repositories/user_repository_impl.ts]: add inside if condition for api

File: packages/database-watermelon/src/dao/user_dao.ts (hunk 5 above)
Comment (Mohan29997) [reply packages/database-watermelon/src/dao/user_dao.ts]: removed

File: packages/database-watermelon/src/dao/user_dao.ts (hunk 6 above)
Comment (Mohan29997) [reply packages/database-watermelon/src/dao/user_dao.ts]: remved it

File: packages/database-watermelon/src/safe_db_call.ts (hunk 4 above)
Comment (Mohan29997) [reply packages/database-watermelon/src/safe_db_call.ts]: safeDbCall(..) make

File: packages/domain/src/usecases/login_check_usecase.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  readonly email?: string
+  readonly password?: string
+
+  constructor(params?: { email: string; password: string }) {
+    super({})
+    this.email = params.email
+    this.password = params.password
+  }
+
+  verify(): boolean {
```
Comment (faiyyazs) [inline packages/domain/src/usecases/login_check_usecase.ts:30]: Check Usecase conditions

File: packages/mobile/src/utils/status.ts [hunk 9]
```diff
@@ -0,0 +1,8 @@
+enum Status {
+  none = 0,
```
Comment (faiyyazs) [inline packages/mobile/src/utils/status.ts:2]: Its redundancy, please remove

File: packages/network-retrofit/src/services/retrofit_service.ts [hunk 10]
```diff
@@ -1,10 +1,11 @@
...
+import { BasePath, BaseService, GET, POST, Response, Headers, Body } from 'ts-retrofit'
+import { UserModel } from 'shared'
 
-@BasePath('/api/')
+@BasePath('')
 class RetrofitService extends BaseService {
-  @POST('login')
-  async login(): Promise<Response<MyLoginResponse>> {
+  @POST('/login')
+  async login(@Body item: UserModel): Promise<Response<MyLoginResponse>> {
```
Comment (faiyyazs) [inline packages/network-retrofit/src/services/retrofit_service.ts:8]: do not use the domain model directly

File: packages/network-retrofit/src/safe_api_call.ts [hunk 11]
```diff
@@ -0,0 +1,10 @@
+async function safeCall<T>(apicall: Promise<T>): Promise<T> {
+  try {
+    const loginResponse = await apicall
```
Comment (faiyyazs) [inline packages/network-retrofit/src/safe_api_call.ts:3]: generic variable

File: packages/network-retrofit/src/model/my_login_response.ts [hunk 12]
```diff
@@ -1,16 +1,13 @@
...
-import { MyFirstModel } from './../../../shared/src/model/myfirstmodel'
+import { UserModel } from 'shared'
 
-export abstract class MyLoginResponse implements BaseLayerDataTransformer<MyLoginResponse, MyFirstModel> {
-  userId: number
-  id: number
-  title: string
-  completed: boolean
+export abstract class MyLoginResponse implements BaseLayerDataTransformer<MyLoginResponse, UserModel> {
+  access_token: any
```
Comment (faiyyazs) [inline packages/network-retrofit/src/model/my_login_response.ts:5]: string type

File: packages/database-watermelon/src/database_adapter.ts [hunk 13]
```diff
@@ -1,20 +1,27 @@
...
   readonly databases: AppDatabase
 
   constructor(params: { databases: AppDatabase }) {
     this.databases = params.databases
   }
-  userCheck(user: User): Promise<boolean> {
-    console.log('database adapter')
-    return this.databases.userDao.checkIfUserLoggedIn({ email: user.email })
+
+  async userDatabaseCall(params?: { email: string; password: string; token: string }): Promise<boolean> {
```
Comment (faiyyazs) [inline packages/database-watermelon/src/database_adapter.ts:12]: insertUser

Comment (Mohan29997) [reply packages/database-watermelon/src/database_adapter.ts]: done

File: packages/domain/src/usecases/login_check_usecase.ts (hunk 8 above)
Comment (Mohan29997) [reply packages/domain/src/usecases/login_check_usecase.ts:31]: add validation

File: packages/network-retrofit/src/services/retrofit_service.ts (hunk 10 above)
Comment (Mohan29997) [reply packages/network-retrofit/src/services/retrofit_service.ts]: done

File: packages/network-retrofit/src/safe_api_call.ts (hunk 11 above)
Comment (Mohan29997) [reply packages/network-retrofit/src/safe_api_call.ts]: done to networkresponse

File: packages/mobile/src/utils/status.ts (hunk 9 above)
Comment (Mohan29997) [reply packages/mobile/src/utils/status.ts]: removed

File: packages/network-retrofit/src/model/my_login_response.ts (hunk 12 above)
Comment (Mohan29997) [reply packages/network-retrofit/src/model/my_login_response.ts:5]: done

File: packages/network-retrofit/src/safe_api_call.ts [hunk 14]
```diff
@@ -0,0 +1,10 @@
+async function safeCall<T>(apicall: Promise<T>): Promise<T> {
```
Comment (faiyyazs) [inline packages/network-retrofit/src/safe_api_call.ts:1]: safeApiCall

File: packages/network-retrofit/src/network_adapter.ts [hunk 15]
```diff
@@ -1,16 +1,17 @@
-import { MyFirstModel } from './../../shared/src/model/myfirstmodel'
 import RetrofitService from './services/retrofit_service'
 import { NetworkPort } from 'data'
+import safecall from './safe_api_call'
+import { UserModel } from 'shared'
 
 class NetowrkAdapter implements NetworkPort {
```
Comment (faiyyazs) [inline packages/network-retrofit/src/network_adapter.ts:6]: rename NetworkAdapter

File: packages/database-watermelon/src/dao/user_dao.ts [hunk 16]
```diff
@@ -1,20 +1,37 @@
...
+          data.access_token = params.token
+        })
+        return userData
+      })
+    )
   }
 
-  async checkIfUserLoggedIn(params: { email: string }): Promise<boolean> {
-    return (await this.getTable().query().fetchCount()) > 0
+  async getRecord(data: { email: string }): Promise<DbUserModel> {
```
Comment (faiyyazs) [inline packages/database-watermelon/src/dao/user_dao.ts:33]: Take null object into consideration

File: packages/database-watermelon/src/dao/dao_base.ts [hunk 17]
```diff
@@ -12,4 +12,5 @@ export abstract class DaoBase<T extends BaseDatabaseModel, P extends Database> {
   abstract delete(record: T): Promise<T>
   abstract insertOrUpdate(record: T): Promise<T>
   abstract deleteAll(): Promise<boolean>
+  abstract getRecord(record: T): Promise<T>
```
Comment (faiyyazs) [inline packages/database-watermelon/src/dao/dao_base.ts:15]: remove it from base

File: packages/database-watermelon/src/dao/base_dao.ts [hunk 18]
```diff
@@ -21,7 +21,9 @@ export abstract class BaseDao<T extends BaseDatabaseModel> extends DaoBase<T, Ap
   deleteAll(): Promise<boolean> {
     throw new Error('Method not implemented.')
   }
-
+  getRecord(record: T): Promise<T> {
```
Comment (faiyyazs) [inline packages/database-watermelon/src/dao/base_dao.ts:24]: remove it from base

Comment (Mohan29997) [reply packages/database-watermelon/src/dao/base_dao.ts:24]: removed

File: packages/network-retrofit/src/safe_api_call.ts (hunk 14 above)
Comment (Mohan29997) [reply packages/network-retrofit/src/safe_api_call.ts:1]: resolved

File: packages/database-watermelon/src/dao/user_dao.ts (hunk 16 above)
Comment (Mohan29997) [reply packages/database-watermelon/src/dao/user_dao.ts]: done

**Review by faiyyazs — APPROVED** [review]
-

**Pull request merged**